### PR TITLE
MCR-2340 add label support for mods classification xslt3 function

### PR DESCRIPTION
**Pull request by rsteph-de**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2340).

### CODE REVIEW

**Review by yagee-de — CHANGES_REQUESTED** [review]
You are working on mycoreclass XML. Your code should go in a more general XSLT file not specifically into mycore-mods.

Then you can chain those functions together and reuse the label function for other non-mods stuff, too.

Function name is "to-mycoreclass" and that should not read as "get-label-in-current-language" ;-)

**Comment by rsteph-de** [comment]
I totally agree with you.
We need a more generic XSLT3 function library for our classification system. This cannot be done by myself alone. We need to collect different usecases and design the API (function names, parameters) well. That should be a topic for our next developers conference.

Until then I would like to keept this functionality here, since this is makes my stylessheets more readable.
old:  `<xsl:value-of select="mcrmods:to-mycoreclass($modsclass/*, 'single')/categories/category/label[@xml:lang=$CurrentLang]/@text" />`
new:  `<xsl:value-of select="mcrmods:to-mycoreclass($modsclass/*, 'label')/@text" />
`
The function still returns the label element of the mycore classification XML and via parent path ("..") you may get the category element and so on.

**Comment by rsteph-de** [comment]
started new xslt function library for classifications

File: mycore-base/src/main/resources/xsl/functions/classification.xsl [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0"?>
+<xsl:stylesheet version="3.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
+  xmlns:xlink="http://www.w3.org/1999/xlink" 
+  xmlns:fn="http://www.w3.org/2005/xpath-functions"
+  xmlns:xs="http://www.w3.org/2001/XMLSchema"
+  xmlns:mcrclass="http://www.mycore.de/xslt/classification"
+  exclude-result-prefixes="fn">
+  <xsl:function name="mcrclass:current-label" as="xs:string?">
+    <xsl:param name="class" as="element()" />
```
Comment (yagee-de) [inline mycore-base/src/main/resources/xsl/functions/classification.xsl:10]: should also work with a specific (root) category to work with complete classifications, too

**Review by yagee-de — CHANGES_REQUESTED** [review]

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 6cfedaabd8249d8f7c07c2b7f37cf87f58ccc526! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6463-pull_request-1141

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 6cfedaabd8249d8f7c07c2b7f37cf87f58ccc526! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6463-pull_request-1141

File: mycore-base/src/main/resources/xsl/functions/classification.xsl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      <xsl:when test="$class//label[@xml:lang=$DefaultLang]">
+        <xsl:sequence select="$class//label[@xml:lang=$DefaultLang]/@text" />
+      </xsl:when>
+      <xsl:when test="$class/@ID">
+        <xsl:sequence select="fn:concat('??', $class/@ID, '@', $CurrentLang, '??')" />
+      </xsl:when>
+      <xsl:otherwise>
+        <xsl:sequence select="()" />
+      </xsl:otherwise>
+    </xsl:choose>
```
Comment (yagee-de) [inline mycore-base/src/main/resources/xsl/functions/classification.xsl]: should use mcrclass:current-label

File: mycore-base/src/main/resources/xsl/functions/classification.xsl [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  xmlns:xlink="http://www.w3.org/1999/xlink" 
+  xmlns:fn="http://www.w3.org/2005/xpath-functions"
+  xmlns:xs="http://www.w3.org/2001/XMLSchema"
+  xmlns:mcrclass="http://www.mycore.de/xslt/classification"
+  exclude-result-prefixes="fn">
+  
+  <xsl:function name="mcrclass:current-label" as="xs:element?">
+    <xsl:param name="class" as="element()" />
+    <xsl:choose>
+      <xsl:when test="$class//label[@xml:lang=$CurrentLang]">
```
Comment (yagee-de) [inline mycore-base/src/main/resources/xsl/functions/classification.xsl]: please replace "//" with "/"

**Review by yagee-de — CHANGES_REQUESTED** [review]

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on b19227ddbe4457e59f5979ced4da718ab7947086! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6500-pull_request-1141

File: mycore-mods/src/main/resources/xsl/functions/mods.xsl [hunk 4]
```diff
@@ -85,7 +85,20 @@
...
+    <xsl:param name="node" as="element()" />
+    <xsl:variable name="class" select="mcrmods:to-mycoreclass($node, 'single')" />
+    <xsl:choose>
+      <xsl:when test="$class/categories/category">
+        <xsl:sequence select="$class/categories/category" />
+      </xsl:when>
+      <xsl:otherwise>
+        <xsl:sequence select="()" />
+      </xsl:otherwise>
+    </xsl:choose>
```
Comment (yagee-de) [inline mycore-mods/src/main/resources/xsl/functions/mods.xsl]: isn't `<xsl:sequence select="$class/categories/category" />` enough here?

File: mycore-base/src/main/resources/xsl/functions/classification.xsl [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  
+  <xsl:function name="mcrclass:current-label" as="element()?">
+    <xsl:param name="class" as="element()" />
+    <xsl:choose>
+      <xsl:when test="$class/label[@xml:lang=$CurrentLang]">
+        <xsl:sequence select="$class/label[@xml:lang=$CurrentLang]" />
+      </xsl:when>
+      <xsl:when test="$class/label[@xml:lang=$DefaultLang]">
+        <xsl:sequence select="$class/label[@xml:lang=$DefaultLang]" />
+      </xsl:when>
```
Comment (yagee-de) [inline mycore-base/src/main/resources/xsl/functions/classification.xsl:17]: should be the `<xsl:otherwise>` case already

**Review by yagee-de — CHANGES_REQUESTED** [review]

File: mycore-base/src/main/resources/xsl/functions/classification.xsl (hunk 5 above)
Comment (rsteph-de) [reply mycore-base/src/main/resources/xsl/functions/classification.xsl:17]: I decided for another default and return the first available label now

File: mycore-mods/src/main/resources/xsl/functions/mods.xsl (hunk 4 above)
Comment (rsteph-de) [reply mycore-mods/src/main/resources/xsl/functions/mods.xsl]: O.K., right, done!

**Review by yagee-de — APPROVED** [review]

**Pull request merged**